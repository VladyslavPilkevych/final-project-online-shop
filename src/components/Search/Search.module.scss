@import '../../assets/styles/global.module.scss';

.searchBox {
  padding: 10px;
  flex-basis: 70%;
}
.searchInput {
  width: 95%;
  border: none;
  outline: none;
  font-size: 14px;
  font-weight: 400;
  line-height: 21px;
  padding: 20px 22px;
  border-radius: 40px;
  color: $dark-grey-text;
  background: $bg-white-color;
}

@media (max-width: 768px) and (min-width: 240px) {
  .searchBox {
    display: none;
  }
  .searchIcon {
    display: none;
  }
}
// @media (max-width: 375px) {
//   .searchBox {
//     display: none;
//   }
//   .searchIcon {
//     display: none;
//   }
// }
