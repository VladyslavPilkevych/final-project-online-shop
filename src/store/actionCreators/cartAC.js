/* eslint-disable no-underscore-dangle */
/* eslint-disable max-len */
/* eslint-disable consistent-return */
import {
  ADD_TO_CART,
  DELETE_FROM_CART,
  CLEAR_CART,
  TOGGLE_CART,
  CREATE_CART,
  GET_CART,
  ON_HANDLE_CART,
  EDIT_CART,
} from '../actions/cartActions';
import * as api from '../../api/api';

export const toggleCart = (value) => ({ type: TOGGLE_CART, payload: value });

export const onHandleCart = (productId) => async (dispatch, getState) => {
  const state = getState();
  if (state.user.user !== null) {
    if (state.cart.dataCart !== null) {
      const data = await api.addToCart(productId);
      dispatch({ type: ADD_TO_CART, payload: data.data.products });
    } else {
      await api.createNewCart(productId);
      const newCart = {
        products: [
          {
            product: productId,
            cartQuantity: 1,
          },
        ],
      };
      dispatch({ type: CREATE_CART, payload: newCart });
    }
  } else {
    localStorage.setItem('cart', JSON.stringify(productId));
  }
};

export const createCart = (productId) => async (dispatch) => {
  const cartData = await api.createNewCart(productId);
  dispatch({ type: CREATE_CART, payload: cartData });
};
export const getCart = (dataCart) => async (dispatch) => {
  const cartData = await api.getCart(dataCart);
  dispatch({ type: GET_CART, payload: cartData.data.products });
};
export const addToCart = (productId) => async (dispatch) => {
  const data = await api.addToCart(productId);
  dispatch({ type: ADD_TO_CART, payload: data });
};
export const editCart = (productId, quantity) => async (dispatch, getState) => {
  const state = getState();
  const tempState = state.cart.dataCart.find((item) => item.product._id === productId).cartQuantity;
  if (quantity > tempState) {
    const cartData = await api.editCart(productId);
    const updatedCart = {
      products: [
        {
          product: productId,
          cartQuantity: quantity,
        },
      ],
    };
    dispatch({ type: EDIT_CART, payload: cartData.data.products });
  } else {
    const cartData = await api.decreaseProductQuantity(productId);
    dispatch({ type: EDIT_CART, payload: cartData.data.products });
  }
};

export const deleteFromCart = (productId) => async (dispatch) => {
  console.log(productId);
  const data = await api.deleteFromCart(productId);
  dispatch({ type: DELETE_FROM_CART, payload: data });
};

export const deleteAllCart = () => async (dispatch) => {
  const data = await api.deleteCart();
  const responseDelete = await data.status;
  if (responseDelete === 200) {
    dispatch({ type: CLEAR_CART, payload: data });
  }
};
