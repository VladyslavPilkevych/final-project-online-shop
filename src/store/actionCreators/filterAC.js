import {
  FILTER_PRODUCTS,
  CLEAR_FILTER,
  CLEAR_FILTER_PRODUCTS,
  FILTER_CATEGORY,
  FILTER_CATEGORY_PRODUCTS,
  ADD_FILTER_COLOR,
  REMOVE_FILTER_COLOR,
  FILTER_BRAND,
  SET_MIN_PRICE_SLIDER_VALUE,
  SET_MAX_PRICE_SLIDER_VALUE,
  NEW_FILTER_PRODUCTS,
  CLEAR_COLOR_FILTER,
  SET_PAGINATION_PAGE,
} from '../actions/filterActions';
import { getFilteredProductsApi } from '../../api/api';

export const filterCategory = (data) => ({ type: FILTER_CATEGORY, payload: data });
export const clearFilter = () => ({ type: CLEAR_FILTER });
export const clearFilterProducts = () => ({ type: CLEAR_FILTER_PRODUCTS });
export const addFilterColor = (data) => ({ type: ADD_FILTER_COLOR, payload: data });
export const removeFilterColor = (data) => ({ type: REMOVE_FILTER_COLOR, payload: data });
export const clearFilterColor = (data) => ({ type: CLEAR_COLOR_FILTER, payload: data });
export const filterBrand = (data) => ({ type: FILTER_BRAND, payload: data });
export const setMinSliderValue = (data) => ({ type: SET_MIN_PRICE_SLIDER_VALUE, payload: data });
export const setMaxSliderValue = (data) => ({ type: SET_MAX_PRICE_SLIDER_VALUE, payload: data });
export const setFilterPaginationPage = (data) => ({ type: SET_PAGINATION_PAGE, payload: data });

export const filterProducts = (data) => async (dispatch) => {
  await getFilteredProductsApi(data)
    .then((rsp) => {
      if (rsp.status === 200) {
        dispatch({ type: FILTER_PRODUCTS, payload: rsp.data.products });
      }
    })
    .catch((err) => {
      console.log(err);
    });
};

export const getCategorieProducts = (url) => async (dispatch) => {
  await getFilteredProductsApi(url)
    .then((rsp) => {
      if (rsp.status === 200) {
        dispatch({ type: FILTER_CATEGORY_PRODUCTS, payload: rsp.data.products });
      }
    })
    .catch((err) => {
      console.error(err);
    });
};

export const newFilterProducts = (data) => async (dispatch) => {
  const dataFilters = [`?categories=${data.categories}`];
  if (data.color.length !== 0) {
    const dataColor = `&color=${data.color.join()}`;
    dataFilters.push(dataColor);
  }
  if (data.name && data.name.length !== 0) {
    const dataBrand = `&name=${data.name.join()}`;
    dataFilters.push(dataBrand);
  }
  if (data.currentPrice && data.currentPrice.min >= 1 && data.currentPrice.max <= 100000) {
    const dataMinPrice = `&minPrice=${data.currentPrice.min}`;
    dataFilters.push(dataMinPrice);
    const dataMaxPrice = `&maxPrice=${data.currentPrice.max}`;
    dataFilters.push(dataMaxPrice);
  }
  await getFilteredProductsApi(dataFilters.join(''))
    .then((rsp) => {
      if (rsp.status === 200) {
        dispatch({ type: NEW_FILTER_PRODUCTS, payload: rsp.data.products });
      }
    })
    .catch((err) => {
      console.log(err);
    });
};