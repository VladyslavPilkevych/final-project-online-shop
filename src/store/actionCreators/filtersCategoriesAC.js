import { TOGGLE_FILTERS_CATEGORIES } from '../actions/filtersCategoriesActions';

const toggleFiltersCategories = (value) => ({ type: TOGGLE_FILTERS_CATEGORIES, payload: value });
export { toggleFiltersCategories };