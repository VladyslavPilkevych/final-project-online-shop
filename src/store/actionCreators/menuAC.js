import { TOGGLE_MENU } from '../actions/menuActions';

export const toggleMenu = (value) => ({ type: TOGGLE_MENU, payload: value });
