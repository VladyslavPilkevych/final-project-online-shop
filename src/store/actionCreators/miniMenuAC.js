import { TOGGLE_MINI_MENU } from '../actions/miniMenuActions';

export const toggleMiniMenu = (value) => ({ type: TOGGLE_MINI_MENU, payload: value });
