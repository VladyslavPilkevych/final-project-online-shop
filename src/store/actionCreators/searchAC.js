import { TOGGLE_SEARCH, PRODUCTS_SEARCH } from '../actions/searchActions';
import { searchProductsApi } from '../../api/api';
import { addNewError } from './productsAC';

export const toggleSearch = (value) => ({ type: TOGGLE_SEARCH, payload: value });

export const searchProducts = (searchPhrases) => async (dispatch) => {
  await searchProductsApi(searchPhrases)
    .then((rsp) => {
      if (rsp.status === 200) {
        dispatch({ type: PRODUCTS_SEARCH, payload: rsp.data });
      }
    })
    .catch((err) => {
      dispatch(addNewError(err));
      // console.log(err);
      // console.log(searchPhrases);
    });
};
