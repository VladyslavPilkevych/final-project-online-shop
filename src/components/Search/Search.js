import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import { toggleSearch, searchProducts } from '../../store/actionCreators/searchAC';
import { ReactComponent as SearchIcon } from '../../assets/icons/Vector.svg';

import styles from './Search.module.scss';

function Search() {
  const isOpenSearch = useSelector((state) => state.search.isOpenSearch);
  const [value, setValue] = useState('');
  const navigate = useNavigate();

  const dispatch = useDispatch();
  if (!isOpenSearch) return null;
  // const closeSearch = () => {
  //   dispatch(toggleSearch(true));
  //   setValue('');
  // };

  const phrase = { query: value };
  const emptyPhraseWithSpace = { query: ' ' };

  const putSearchedProducts = () => {
    if (value === '') {
      dispatch(searchProducts(emptyPhraseWithSpace));
    } else {
      dispatch(searchProducts(phrase));
    }
  };

  return (
    <div className={styles.searchBox}>
      <form
        onSubmit={(e) => {
          e.preventDefault();
          putSearchedProducts();
          localStorage.setItem('phrase', JSON.stringify(phrase));
          navigate({ pathname: '/products/search' });
          setValue('');
        }}
      >
        <input
          className={styles.searchInput}
          type="text"
          placeholder="Search entiere store here..."
          value={value}
          onChange={(e) => {
            setValue(e.target.value);
          }}
        />
      </form>
    </div>
  );
}

export default Search;
