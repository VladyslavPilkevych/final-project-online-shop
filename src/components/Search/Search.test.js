/* eslint-disable arrow-body-style */
/* eslint-disable react/function-component-definition */
/* eslint-disable react/button-has-type */
/* eslint-disable react/react-in-jsx-scope */
import {
  getByDisplayValue, getByPlaceholderText, render, screen,
} from '@testing-library/react';
import { Provider, useDispatch } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import store from '../../store';
import Search from './Search';
import { toggleSearch, searchProducts } from '../../store/actionCreators/searchAC';

const Component = () => {
  return (
    <BrowserRouter>
      <Provider store={store}>
        <Search />
        <OpenButton />
        <CloseButton />
        <SearchButton />
      </Provider>
    </BrowserRouter>
  );
};

const OpenButton = () => {
  const dispatch = useDispatch();
  const toggleOpen = () => {
    dispatch(toggleSearch(true));
  };

  return (
    <button onClick={toggleOpen}>Open</button>
  );
};

const CloseButton = () => {
  const dispatch = useDispatch();
  const toggleClose = () => {
    dispatch(toggleSearch(false));
  };

  return (
    <button onClick={toggleClose}>Close</button>
  );
};

const SearchButton = () => {
  const dispatch = useDispatch();
  const searchForProduct = () => {
    dispatch(searchProducts('Apple'));
  };
  return (
    <button onClick={searchForProduct}>Search</button>
  );
};

describe('Snapshot search with button', () => {
  test('Should snapshot search with button', () => {
    const { asFragment } = render(<Component />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Search should open', () => {
  test('Search is opened', () => {
    const { asFragment, getByText } = render(<Component />);
    const button = getByText('Open');
    button.click();
    expect(screen.getByPlaceholderText('Search entiere store here...')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Search should close', () => {
  test('Search is closed', () => {
    const { asFragment, getByText, queryByPlaceholderText } = render(<Component />);
    const button = getByText('Close');
    button.click();
    expect(queryByPlaceholderText('Search entiere store here...')).not.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('Search input should work', () => {
  test('Search value is input value', () => {
    const { getByText } = render(<Component />);
    const button = getByText('Open');
    button.click();
    userEvent.type(screen.getByPlaceholderText('Search entiere store here...'), 'Apple');
    expect(screen.getByPlaceholderText('Search entiere store here...').value).toBe('Apple');
  });

  test('Search should find products, by input value', () => {
    const { getByText } = render(<Component />);
    const openButton = getByText('Open');
    const searchButton = getByText('Search');
    openButton.click();
    userEvent.type(screen.getByPlaceholderText('Search entiere store here...'), 'Apple');
    searchButton.click();
  });
});
