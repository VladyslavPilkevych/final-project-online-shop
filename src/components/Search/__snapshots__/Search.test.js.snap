// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should close Search is closed 1`] = `
<DocumentFragment>
  <button>
    Open
  </button>
  <button>
    Close
  </button>
  <button>
    Search
  </button>
</DocumentFragment>
`;

exports[`Search should open Search is opened 1`] = `
<DocumentFragment>
  <div
    class="searchBox"
  >
    <form>
      <input
        class="searchInput"
        placeholder="Search entiere store here..."
        type="text"
        value=""
      />
    </form>
  </div>
  <button>
    Open
  </button>
  <button>
    Close
  </button>
  <button>
    Search
  </button>
</DocumentFragment>
`;

exports[`Snapshot search with button Should snapshot search with button 1`] = `
<DocumentFragment>
  <button>
    Open
  </button>
  <button>
    Close
  </button>
  <button>
    Search
  </button>
</DocumentFragment>
`;
